test/functional/wallet_coinselection.py

```diff
@@ -0,0 +1,261 @@
+
+    def test_one_big_and_many_small_coins(self):
+        w0 = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+        self.nodes[0].createwallet(wallet_name="test_one_big_and_many_small_coins")
+        w1 = self.nodes[0].get_wallet_rpc("test_one_big_and_many_small_coins")
+
+        # Create a bunch of small coins
+        for i in range(100):
+            w0.sendtoaddress(w1.getnewaddress("", "legacy"), 0.05)
+
+        # We will target a solution with 20 coins that will add up exactly to 1btc.
+        # Add a coin that will cover all the fees and exactly match the target.
+        # TODO: fix that knapsack targets change even if there is no change needed
+        # (148*20+42+31)*350+5000000 = 6061550
+        w0.sendtoaddress(w1.getnewaddress("", "legacy"), 0.0606155)
+        self.generate(self.nodes[0], 6)
+
+        fee_rate = 350
+        dummyaddress = w0.getnewaddress()
+
+        def select_coins(big_coin, target, iterations=100):
+            txid = w0.sendtoaddress(w1.getnewaddress("", "bech32"), big_coin)
+            self.generate(self.nodes[0], 6)
+
+            fee_total = 0
+            max_fee = 0
+            for _ in range(iterations):
+                result = w1.fundrawtransaction(w1.createrawtransaction([], {dummyaddress: target}), {"fee_rate": fee_rate})
+                fee_total += result['fee']
+                max_fee = max(max_fee, result['fee'])
+
+            # lock big coin
+            w1.lockunspent(False, [tx for tx in w1.listunspent() if tx['txid'] == txid])
+            return fee_total, max_fee
+
+        # in all the scenarios below the best solution would be to take one big coin and create a change output
+        best_fee = (42 + 68 + 31) * fee_rate / COIN
+
+        # 0) no exact match is possible, the big coin is closest to target + MIN_CHANGE,
+        # knapsack always finds the best solution
+        fee_total, max_fee = select_coins(big_coin=1.01051, target=1.00000001, iterations=100)
+        assert_approx(max_fee, best_fee)
+        assert_approx(fee_total, best_fee * 100)
+
+        # 1) knapsack latches to the exact match solution and can't find the best solution
+        # SRD can find a solution better than knapsack if draws the big coin in the first 21 tries
+        # chances for SRD to find it are roughly 2*(21/102) = ±40%.
+        # Chances are doubled due to coinselection run twice with and without APS enabled
+        fee_total, max_fee = select_coins(big_coin=1.01051, target=1.0, iterations=100)
+        knapsack_fee = (42 + 31 + 148*20) * fee_rate / COIN
+        assert_approx(max_fee, knapsack_fee)
+        # total fee is lower due to lucky SRD solutions
+        assert_greater_than(knapsack_fee * 100, fee_total)
+        assert_greater_than(fee_total, best_fee * 100)
+
+        # 2) no exact match, effective value of the big coin is smaller than target + MIN_CHANGE
+        # BnB can find the same solution from previous test but dropping the change output to fees
```

Comment:
```suggestion
        # BnB will find the same solution from previous test but dropping the change output to fees
```
BnB is deterministic, so if it "can" I would expect that it does? Does "change output" mean "remainder" here?—Is this the BnB without an upper bound or the variant we currently use?

Reply:
Any BnB, since we're within the cost of change bound. I'll retouch the comment